### DISCUSSION TITLE
Some inconsistencies observed with HarfBuzz output for Gujarati fonts

### DISCUSSION THREAD

**Discussion by sid-wr** [harfbuzz/harfbuzz#4281] [General] [answered]
Hi team,

I tried using **hb-view** tool for some **Gujarati** text & some broken clusters can be observed on checking the generated output.

The input text is ```-u "U+0AB8U+0ACDU+0AB5U+0ABEU+0AA4U+0A82U+0AA4U+0ACDU+0AB0U+0ACDU+0AAF"```

The expected output is ```સ્વાતંત્ર્ય```

I tried this for 20 different fonts & the Gujarati cluster is getting broken for the following 8 fonts:
1. Adobe Gujarati
2. Amdavadi Unicode
3. Bhavnagar Unicode
4. Bhuj Unicode
5. Rajkot Unicode
6. Surat Unicode
7. Vadodara Unicode
8. Vijaya Unicode

For the rest 12 fonts, the correct expected cluster is generated:
1. Skolar Gujarati
2. Gujarati Sangam
3. Kohinoor Gujarati
4. Nirmala UI
5. NotoSansGujaratiUI
6. NotoSerifGujarati
7. Ekatra
8. Anek Gujarati
9. Shruti
10. Padmaa
11. Rekha
12. Aakar

Also, I tried some other licensed third-party shaper software & this cluster is not getting broken for any of these 8 fonts at their end.
So can you please check whether the issue is at your end or at the respective fonts side ?

Have attached a combined screenshot of all outputs which I created for easy comparison:
<img width="517" alt="Comparison of outputs with different Gujarati fonts" src="https://github.com/harfbuzz/harfbuzz/assets/136314541/28b562f0-21b1-44a0-847f-ed9186de6561">

Please help
Thanks

**Comment by behdad** [comment]
> I tried this for 20 different fonts & the Gujarati cluster is getting broken for the following 8 fonts:

These are probably indic1 fonts...  No idea why broken. Just speculating.

**Reply by behdad** [reply]
I think this is because we look into the font for base-finding.

cc @jfkthame

**Comment by behdad** [comment]
> These are probably indic1 fonts... No idea why broken. Just speculating.

That's not the reason.

With this sequence (ta, virama, ra, virama, ya):
```
build/util/hb-shape amdavad-unicode.ttf -u "0AA4 0ACD 0AB0 0ACD 0AAF" --trace
```
I see `ta,virama` form half-form glyph150 in the `half` feature. Then in `pres` feature that forms glyph181 with the ra. But the font has a half-form for glyph181,virama in the `half` feature, which has been long processed by now. So I'm not sure how this is supposed to work.

**Reply by behdad** [reply]
Ah. I think it *is* a base-finding failure. We set the base to ya. Whereas in indic1 it is expected to set the base to ta.

**Reply by behdad** [reply]
Humm. I'm not sure now. Changing the base to 0 manually in the code doesn't fix the rendering.

**Comment by behdad** [comment]
> Ah. I think it _is_ a base-finding failure. We set the base to ya. Whereas in indic1 it is expected to set the base to ta.

@jfkthame Should we hardcode the consonant positions for indic1?

**Comment by behdad** [comment]
What's happening is that ra,virama is forming a ligature in `blwf` that is necessary for the whole conjunct to work. But we only apply blwf to post-base glyphs. And since we find base to be at 4, we don't apply it.

**Comment by sid-wr** [comment]
Hi,
Should I log a new bug for this?

**Reply by behdad** [reply, marked as answer]
I just did. Thanks.
